tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/balancing/ElasticQueueTest.java

```diff
@@ -0,0 +1,697 @@
+import org.apache.activemq.artemis.spi.core.protocol.RemotingConnection;
+import org.apache.activemq.artemis.spi.core.security.ActiveMQSecurityManager5;
+import org.apache.activemq.artemis.spi.core.security.jaas.RolePrincipal;
+import org.apache.activemq.artemis.spi.core.security.jaas.UserPrincipal;
+import org.apache.activemq.artemis.tests.integration.management.ManagementControlHelper;
+import org.apache.activemq.artemis.tests.util.ActiveMQTestBase;
+import org.apache.activemq.artemis.utils.Wait;
+import org.apache.qpid.jms.JmsConnection;
+import org.apache.qpid.jms.JmsConnectionFactory;
+import org.apache.qpid.jms.JmsConnectionListener;
+import org.apache.qpid.jms.message.JmsInboundMessageDispatch;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ElasticQueueTest extends ActiveMQTestBase {
+
+   static final String qName = "EQ";
+   static final SimpleString qNameSimple = SimpleString.toSimpleString(qName);
+
+   final int base_port = 61616;
+   final int opTimeoutMillis = 2000;
+   final Stack<EmbeddedActiveMQ> nodes = new Stack<>();
+   private final String balancerConfigName = "role_name_sharder";
+
+   String urlForNodes(Stack<EmbeddedActiveMQ> nodes) {
+      StringBuilder builder = new StringBuilder("failover:(");
+      int port_start = base_port;
+      for (EmbeddedActiveMQ activeMQ : nodes) {
+         if (port_start != base_port) {
+            builder.append(",");
+         }
+         builder.append("amqp://localhost:" + (port_start++));
+      }
+      // fast reconnect, randomize to get to all brokers and timeout sends that block on no credit
+      builder.append(")?failover.maxReconnectDelay=2000&failover.randomize=true&jms.sendTimeout=" + opTimeoutMillis);
+      return builder.toString();
+   }
+
+   // allow tracking of failover reconnects
+   class ConnectionListener implements JmsConnectionListener {
+
+      AtomicInteger connectionCount;
+
+      ConnectionListener(AtomicInteger connectionCount) {
+         this.connectionCount = connectionCount;
+      }
+
+      @Override
+      public void onConnectionEstablished(URI uri) {
+      }
+
+      @Override
+      public void onConnectionFailure(Throwable throwable) {
+      }
+
+      @Override
+      public void onConnectionInterrupted(URI uri) {
+      }
+
+      @Override
+      public void onConnectionRestored(URI uri) {
+         connectionCount.incrementAndGet();
+      }
+
+      @Override
+      public void onInboundMessage(JmsInboundMessageDispatch jmsInboundMessageDispatch) {
+      }
+
+      @Override
+      public void onSessionClosed(Session session, Throwable throwable) {
+      }
+
+      @Override
+      public void onConsumerClosed(MessageConsumer messageConsumer, Throwable throwable) {
+      }
+
+      @Override
+      public void onProducerClosed(MessageProducer messageProducer, Throwable throwable) {
+      }
+   }
+
+   // slow consumer
+   class EQConsumer extends Thread {
```

Comment:
Probably nicer to use an executor which can then be cleaned up afterwards

Reply:
agree.